### PR TITLE
[EPLB] Add EPLB support for hunyuan_v1

### PR DESCRIPTION
**Pull request by 666even666**
## Purpose

This pull request add EPLB support for hunyuan_v1 model, which helps to improve the overall thoughput during LLM Serving.

https://github.com/vllm-project/vllm/issues/20468

## Test Plan
```
import json
import os
import argparse
from vllm import LLM, SamplingParams

prompt = "Where is the capital of Russia? Please provide a brief explanation."

RESULT_FILE = "eplb_test_output.json"

sampling_params = SamplingParams(
    temperature=0.0,
    top_p=1.0,
    top_k=1,
    max_tokens=100
)

def run_inference(enable_eplb: bool, num_redundant_experts: int = 0):
    print(f"Running inference with EPLB={enable_eplb}, redundant experts={num_redundant_experts}")

   llm = LLM(
        model="tencent/Hunyuan-A13B-Instruct-FP8",
        tensor_parallel_size=2,
        enable_expert_parallel=True,
        enable_eplb=enable_eplb,
        num_redundant_experts=num_redundant_experts if enable_eplb else 0,
        eplb_window_size=1000,
        eplb_step_interval=100,
        eplb_log_balancedness=True,
        enforce_eager=True,
        trust_remote_code=True
    )

    result = llm.generate([prompt], sampling_params)
    output_text = result[0].outputs[0].text.strip()

    print("Output:")
    print(output_text)
    print("-" * 50)

    return output_text

def save_result(key: str, value: list):
    if os.path.exists(RESULT_FILE):
        with open(RESULT_FILE, "r") as f:
            results = json.load(f)
    else:
        results = {}

    results[key] = value

    with open(RESULT_FILE, "w") as f:
        json.dump(results, f, indent=2)

    print(f"Output saved to {RESULT_FILE}")

def load_results():
    if os.path.exists(RESULT_FILE):
        with open(RESULT_FILE, "r") as f:
            return json.load(f)
    return {}

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", type=str, choices=["eplb", "normal", "compare"], required=True)
    args = parser.parse_args()

    if args.mode == "eplb":
        outputs = run_inference(enable_eplb=True, num_redundant_experts=32)
        save_result("eplb", outputs)
    elif args.mode == "normal":
        outputs = run_inference(enable_eplb=False)
        save_result("normal", outputs)
```

execute
```
python eplb_test.py --mode eplb
```
## Test Result

after
rearrange

## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by 666even666** [comment]
@abmfy This PR adds EPLB support to hunyuan.  I used tencent/Hunyuan-A13B-Instruct-FP8 for testing and had to skip w13_input_scale and w2_input_scale in get_expert_weights as they are scalars and cannot be reshaped to (self.local_num_experts, -1). Let me know if you have other suggestions :)

**Comment by abmfy** [comment]
Thanks for the contribution!  

I haven’t had time to run an accuracy test yet, but you shouldn’t skip the scale tensors -- otherwise, the weights and scales of the physical experts will become misaligned after a rearrangement.  

If I recall correctly, the scale tensor looks like this:  

```python
w13_input_scale = torch.nn.Parameter(
    torch.ones(num_experts, dtype=torch.float32),
    requires_grad=False,
)
```

So it’s still a tensor with a dimension of `num_local_experts`.  

Could you attach an accuracy test on GSM8K after fixing this issue?

**Comment by 666even666** [comment]
Thanks for the review @abmfy!  I do see that w13_input_scale is set to torch.ones(num_experts, dtype=torch.float32) in fp8.py.   However when I print out tensor shapes in layer.py it shows that w13_input_scale is a scalar 
scalar  and if I don't skip w13_input_scale, I always end up with an error complaining that scalar cannot be reshaped to (self.local_num_experts, -1).  

I will try to debug this issue. Let me know if you have any insights in the meantime :)

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -1344,6 +1344,9 @@ def get_expert_weights(self) -> Iterable[torch.Tensor]:
         # with shape (num_logical_experts,), not an expert weight.
         NON_EXPERT_WEIGHTS = {
             "e_score_correction_bias",
+            # skip input scales as they are scalers
+            "w13_input_scale",
+            "w2_input_scale",
```
Comment (abmfy) [inline vllm/model_executor/layers/fused_moe/layer.py]: Hi @666even666, it looks like in [fp8.py#L797-L800](https://github.com/vllm-project/vllm/blob/8ef6b8a38cc3b69e55b0bb60d5442c4efb1988de/vllm/model_executor/layers/quantization/fp8.py#L797-L800) the `input_scale`s are converted into a single-element tensor when the activation scheme is `static`, which is the case for Hunyuan-FP8 models.  

In this case, these scales don’t need to be transferred since they’re shared across the whole layer. However, we shouldn’t skip them unconditionally, as that might break models using the `dynamic` activation scheme. I think we should instead add a check to skip only one-element tensors (with shape `[]`). What do you think?

**Review by abmfy — CHANGES_REQUESTED** [review]

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (666even666) [reply vllm/model_executor/layers/fused_moe/layer.py]: Thanks a lot for spotting it @abmfy ! I have added the scalar check as suggested.  BTW is there a recommended way to run accuracy test on GSM8K (e.g. existing test script/process) ?

Comment (abmfy) [reply vllm/model_executor/layers/fused_moe/layer.py]: I've been using this one and compare the numbers before/after enabling EPLB:
```bash
PORT=${PORT:-8080}
MODEL="MODEL_NAME_HERE"

OPENAI_API_KEY=EMPTY lm_eval \
    --model local-completions \
    --model_args "base_url=http://localhost:$PORT/v1/completions,model=$MODEL,num_concurrent=4096,trust_remote_code=True" \
    --tasks gsm8k \
    --output_path results/gsm8k.json
```

You'll need to `uv pip install lm_eval[api]` first

**Comment by 666even666** [comment]
Hi @abmfy 

Here is the accuracy test  result on GSM8K 

Without EPLB
hunyuaneplb

With EPLB
hunyuan

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (666even666) [reply vllm/model_executor/layers/fused_moe/layer.py]: Thank you so much for the suggestion! I have attached test results below :)

**Review by abmfy — APPROVED** [review]
LGTM! Thanks for the contribution 🎉

One last thing: have you confirmed that the balancedness score increases with `log_balancedness: true` set in `EplbConfig`? If so, I think we're good to go and can merge once CI passes.  

cc @simon-mo Could you please approve to trigger CI? Thank you!

**Comment by 666even666** [comment]
@abmfy Thanks for the review! From the screenshot I took, I do see the balancedness score increased but not in a monotonic way. Do you think this is expected? (I am using ep size of 2)
rearrange
test

**Comment by abmfy** [comment]
> @abmfy Thanks for the review! From the screenshot I took, I do see the balancedness score increased but not in a monotonic way. Do you think this is expected? (I am using ep size of 2) rearrange test

The figures here aren’t very informative since we only used a single prompt with EP=2. Ideally, we should run benchmarks with larger batches and a larger-scale EP setup; otherwise, the improvements brought by EPLB may not be very apparent. 

That said, I believe the model adaptation is working well, so we’re good to proceed.

Thanks again for the contribution! 🎉

**Comment by 666even666** [comment]
Ahh I see. That makes a lot of sense. Thanks a lot @abmfy for your help and support!!

**Comment by 666even666** [comment]
Looks like we still need another review to merge the pr. @abmfy Do you know someone who can help review the change :)

**Comment by abmfy** [comment]
@simon-mo Could you please take a look and help merge this if it doesn’t affect other parts of the codebase? Thanks a lot!

**Review by simon-mo — APPROVED** [review]
